sdk/ai/Azure.AI.Projects/src/Custom/AIProjectClient.cs

```diff
@@ -39,61 +77,162 @@ public AIProjectClient(string connectionString, TokenCredential credential, AIPr
-            ConnectionType connectionType = useServerlessConnection ? ConnectionType.Serverless : ConnectionType.AzureAIServices;
+            // Check if the connection details are already cached (read lock).
+            ConnectionResponse connection = null;
+            _connectionCacheLock.EnterReadLock();
+            try
+            {
+                _connectionCache.TryGetValue(connectionType, out connection);
+            }
+            finally
+            {
+                _connectionCacheLock.ExitReadLock();
+            }
 
-            ConnectionResponse connection = connectionsClient.GetDefaultConnection(connectionType, true);
+            // If not in cache, acquire a write lock to populate it.
+            if (connection == null)
+            {
+                _connectionCacheLock.EnterWriteLock();
+                try
+                {
+                    // Double-check in case another thread already added it.
+                    if (!_connectionCache.TryGetValue(connectionType, out connection))
+                    {
+                        connection = _connectionsClient.GetDefaultConnection(connectionType, true);
+                        _connectionCache[connectionType] = connection;
+                    }
+                }
+                finally
+                {
+                    _connectionCacheLock.ExitWriteLock();
+                }
+            }
 
+            // If the connection uses API key auth, validate and add if needed.
             if (connection.Properties is ConnectionPropertiesApiKeyAuth apiKeyAuthProperties)
             {
                 if (string.IsNullOrWhiteSpace(apiKeyAuthProperties.Target))
                 {
-                    throw new ArgumentException("The API key authentication target URI is missing or invalid.");
+                    throw new ArgumentException(
+                        $"The API key authentication target URI is missing or invalid for {connectionId}.");
                 }
 
-                if (!Uri.TryCreate(apiKeyAuthProperties.Target, UriKind.Absolute, out var endpoint))
+                if (apiKeyAuthProperties.Credentials == null
+                    || string.IsNullOrWhiteSpace(apiKeyAuthProperties.Credentials.Key))
                 {
-                    throw new UriFormatException("Invalid URI format in API key authentication target.");
+                    throw new ArgumentException($"The API key is missing or invalid for {connectionId}.");
                 }
 
-                var credential = new AzureKeyCredential(apiKeyAuthProperties.Credentials.Key);
-                if (!useServerlessConnection)
+                // Build the new connection object.
+                var newConnection = new ClientConnection(connectionId, apiKeyAuthProperties.Target, apiKeyAuthProperties.Credentials.Key);
+
+                // Now we need to re-check and possibly add to Connections under a write lock.
+                _connectionsLock.EnterUpgradeableReadLock();
```

Comment:
Yeah, all the synchronization code should be either in SCM or at lease encapsulated in some internal type in this package. We don't want this complex code to be intermixed with the rest of the project implementation.